### PR TITLE
MemberRef "$R" and static import support added.

### PR DESCRIPTION
**Pull request by sormuras**
#317 and #73 solved?

* ~~README.md needs some "$R"-love, read paragraph.~~ https://github.com/sormuras/javapoet/tree/ref#r-for-references-and-static-imports
* What about references to `MethodSpec`, `FieldSpec` and `TypeSpec.enumConstants`? Or are they covered by `$N` already?

### CODE REVIEW

File: src/main/java/com/squareup/javapoet/MemberRef.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    Util.checkNotNull(types, "types == null");
+    TypeName type = TypeName.get(element.getEnclosingElement().asType());
+    String name = element.getSimpleName().toString();
+    return new MemberRef(type, name, someObject, TypeName.list(types));
+  }
+
+  public final TypeName type;
+  public final List<TypeName> arguments;
+  public final String name;
+  public final Object someObject;
```
Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/MemberRef.java]: I don't quite understand what this is.

File: src/main/java/com/squareup/javapoet/MemberRef.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    Util.checkNotNull(element, "element == null");
+    Util.checkNotNull(someObject, "someObject == null");
+    Util.checkNotNull(types, "types == null");
+    TypeName type = TypeName.get(element.getEnclosingElement().asType());
+    String name = element.getSimpleName().toString();
+    return new MemberRef(type, name, someObject, TypeName.list(types));
+  }
+
+  public final TypeName type;
+  public final List<TypeName> arguments;
```
Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/MemberRef.java]: Ooh, are these type parameters? Hmmm

**Comment by swankjesse** [comment]
Do you mind sitting on this for a while until I have time to figure out how this is most useful?

The thing that's really tricky for me to reason about is how these things work in practice. Typically you need to dereference either a static thing `Roshambo.ROCK` or an instance thing `roshambo.name()`.

What we have here _works_ but I really want to explore the consequences of the design, including how it'll work for static imports, and chained references.

**Comment by sormuras** [comment]
Rebased and I'm totally fine with exploring.

See #317 for some discussion around the topic.
And https://github.com/sormuras/javapoet/tree/ref#r-for-references for usage examples.

File: src/main/java/com/squareup/javapoet/MemberRef.java (hunk 2 above)
Comment (sormuras) [inline src/main/java/com/squareup/javapoet/MemberRef.java]: https://github.com/square/javapoet/issues/317#issuecomment-148513048

File: src/main/java/com/squareup/javapoet/MemberRef.java (hunk 1 above)
Comment (sormuras) [inline src/main/java/com/squareup/javapoet/MemberRef.java]: What is the line, you're referring to?

If it's line 73 `public final Object someObject;` then it's the call site, the target, the instance that the method is called on.

File: src/test/java/com/squareup/javapoet/MemberRefTest.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+  @Test public void readmeExampleWithMethodChaining() throws Exception {
+    // get members
+    Method cbBuilder = CodeBlock.class.getMethod("builder"); // static
+    Method cbAdd = CodeBlock.Builder.class.getMethod("add", String.class, Object[].class);
+    Method cbBuild = CodeBlock.Builder.class.getMethod("build");
+    // create refs explicitly
+    TypeName tnCodeBlock = TypeName.get(CodeBlock.class);
+    MemberRef mrBuilder = MemberRef.get(cbBuilder, tnCodeBlock, new Type[0]);
+    MemberRef mrAdd = MemberRef.get(cbAdd, "");
+    MemberRef mrBuild = MemberRef.get(cbBuild, "");
```
Comment (sormuras) [inline src/test/java/com/squareup/javapoet/MemberRefTest.java]: Overloaded `get` too much. Having to pass an empty array is not ... nice.

Will introduce more specific getters, like `getStaticMemberRef`, `getInstanceMemberRef`

File: src/test/java/com/squareup/javapoet/MemberRefTest.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    // wire up
+    CodeBlock block = CodeBlock.builder()
+        .add("$R()$R(\"$$R\", $R)$R()", mrBuilder, mrAdd, Thread.State.NEW, mrBuild)
```
Comment (sormuras) [inline src/test/java/com/squareup/javapoet/MemberRefTest.java]: Method arguments are part of the format string. Makes the reference instance re-usable for various arguments.
The chaining `.`s are part the `$R` rendering with the empty literal `""` as a target. Needs some more thoughts and work.

**Comment by sormuras** [comment]
#73 ... I'd expect that `JavaFile` takes an user-controlled list of member refs to import, right? Meaning, no auto-magic import guessing here.

**Comment by sormuras** [comment]
Do you mind reviewing again? :beers: 

The implementation of #73 is a initial one - but works nicely at the test cases. Needs a ton o'tests more, though.

**Comment by sormuras** [comment]
Rebased to javapoet/master.

**Comment by sormuras** [comment]
In this context, [Java 8 method references](https://docs.oracle.com/javase/tutorial/java/javaOO/methodreferences.html) need some love/thoughts.

Without thinking in detail, you can't use `Class::member` as an input for `MemberRef.get(...)` as the actual target is hidden by the lambda construct.

**Comment by sormuras** [comment]
Rebased on 1.4.0 release.

**Comment by sormuras** [comment]
@swankjesse The MemberRef property `Set<Modifier> modifiers` could easily be reduced to a `boolean isStatic` flag. At the moment, only that information is needed and used by the emitting logic.

**Comment by sormuras** [comment]
Rebased.

**Comment by sormuras** [comment]
I'l remove the "static import" support from this PR within the next days. It`s only working via manual tagging and I have an auto-import in my mind, and already have-coded.

**Comment by sormuras** [comment]
@Deprecated - use #388 and/or following instead.

**Pull request closed without merging**